src/main/java/io/github/thebusybiscuit/slimefun4/implementation/accessibility/PlayerAccessManagerImpl.java

```diff
@@ -0,0 +1,247 @@
+        loadFromJsonElement(element);
+    }
+
+    @Nonnull
+    @Override
+    public AccessLevel getAccessLevel(@Nonnull final Player player) {
+        return getAccessLevel(player.getUniqueId());
+    }
+
+    @Nonnull
+    @Override
+    public AccessLevel getAccessLevel(@Nonnull final UUID player) {
+        return getAccessData(UUID.class).map(accessData -> accessData.getAccessLevel(player))
+            .orElse(ConcreteAccessLevel.NONE);
+    }
+
+    @Nonnull
+    @Override
+    public <T> Optional<AccessData<T>> getAccessData(@Nonnull final Class<T> object) {
+        final AccessData<?> raw = dataMap.get(object);
+        if (raw != null) {
+            return Optional.of(((AccessData<T>) raw));
+        }
+        for (Map.Entry<Class<?>, AccessData<?>> entry : dataMap.entrySet()) {
+            if (object.isAssignableFrom(entry.getKey())) {
+                return Optional.of((AccessData<T>) entry.getValue());
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Nonnull
+    @Override
+    public <T> AccessData<T> getOrRegisterAccessData(@Nonnull final Class<T> clazz,
+                                                     @Nonnull final AccessData<T> def) {
+        return getAccessData(clazz).orElseGet(() -> {
+            dataMap.put(clazz, def);
+            return def;
+        });
+    }
+
+    @Override
+    public boolean hasDataFor(@Nonnull final Object object) {
+        if (object instanceof Player) {
+            return getAccessData(Player.class).map(data -> data.hasDataFor((Player) object))
+                .orElse(false);
+        } else if (object instanceof UUID) {
+            return getAccessData(UUID.class).map(data -> data.hasDataFor((UUID) object))
+                .orElse(false);
+        }
+        final Optional<? extends AccessData> optional = getAccessData(object.getClass());
+        return optional.map(accessData -> accessData.hasDataFor(object)).orElse(false);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return dataMap.isEmpty();
+    }
+
+    @Override
+    public void reset() {
+        this.dataMap.clear();
+    }
+
+    @Override
+    public void load(@Nonnull NamespacedKey namespace, @Nonnull ItemStack itemStack) {
+        reset();
+        if (!SlimefunPlugin.getMinecraftVersion().isAtLeast(MinecraftVersion.MINECRAFT_1_14)) {
+            // This feature was introduced post 1.13. Cannot load newer data on an older version
+            return;
+        }
+        final ItemMeta meta = itemStack.getItemMeta();
+        if (meta != null) {
+            load(namespace, meta);
+        }
+
+    }
+
+    @Override
+    public void load(@Nonnull final NamespacedKey namespace, @Nonnull final Block block) {
+        reset();
+        if (!SlimefunPlugin.getMinecraftVersion().isAtLeast(MinecraftVersion.MINECRAFT_1_14)) {
+            return;
+        }
+        final BlockState state = block.getState();
```

Comment:
#getState() creates a memory and performance heavy snapshot, use PaperLib for this.